subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c

```diff
@@ -94,45 +111,74 @@ static int put_name_r(struct lwm2m_output_context *out, struct lwm2m_obj_path *p
 	return 0;
 }
 
-static int put_begin_oi(struct lwm2m_output_context *out, struct lwm2m_obj_path *path)
+/* TODO: Strictly speaking this fxn is unnecessary but used for error checking, remove later */
+static int put_begin(struct lwm2m_output_context *out, struct lwm2m_obj_path *path)
 {
 	struct cbor_out_formatter_data *fd = LWM2M_OFD_CBOR(out);
+	int ret;
 
-	if (fd->input.rec_cnt >= CONFIG_LWM2M_RW_SENML_CBOR_RECORDS) {
-		__ASSERT_PRINT("CONFIG_LWM2M_RW_SENML_CBOR_RECORDS too small");
-		return -ENOMEM;
+	ret = put_basename(out, path);
+	if (ret < 0) {
+		__ASSERT_NO_MSG(false);
+		return ret;
 	}
 
-	/* Basename is the path a requester requests */
-	int len = lwm2m_path2string(fd->bnames[fd->input.rec_cnt], fd->bname_sz, path,
-				    LWM2M_PATH_LEVEL_OBJECT_INST);
+	struct record *record = GET_CBOR_FD_REC(fd);
 
-	if ((len < sizeof("/0/0/") - 1) || (len > sizeof("/65535/65535/") - 1)) {
+	if ((record->bn.hndl.len < sizeof("/0") - 1) ||
+	    (record->bn.hndl.len > sizeof("/65535") - 1)) {
 		__ASSERT_NO_MSG(false);
 		return -EINVAL;
 	}
 
-	/* Get the first record which will hold the basename */
-	struct record *record = GET_CBOR_FD_REC(fd);
+	/* Does not advance the record count, waiting for data */
+	return 0;
+}
 
-	record->bn.hndl.value = fd->bnames[fd->input.rec_cnt];
-	record->bn.hndl.len = len;
-	record->bn_prsnt = 1;
+static int put_end(struct lwm2m_output_context *out, struct lwm2m_obj_path *path)
+{
+	size_t len;
 
-	/* Does not advance the record count, name next */
-	return 0;
+	bool ret =
+		cbor_encode_lwm2m_senml(CPKT_BUF_W_REGION(out->out_cpkt),
+					(struct lwm2m_senml *)&(LWM2M_OFD_CBOR(out)->input), &len);
+
+	if (ret) {
+		out->out_cpkt->offset += len;
+	}
+
+	return ret ? len : -EINVAL;
 }
 
-static int put_begin_r(struct lwm2m_output_context *out, struct lwm2m_obj_path *path)
+static int put_begin_oi(struct lwm2m_output_context *out, struct lwm2m_obj_path *path)
 {
 	struct cbor_out_formatter_data *fd = LWM2M_OFD_CBOR(out);
+	int ret;
+	uint8_t tmp = path->level;
 
-	if (fd->input.rec_cnt >= CONFIG_LWM2M_RW_SENML_CBOR_RECORDS) {
-		__ASSERT_PRINT("CONFIG_LWM2M_RW_SENML_CBOR_RECORDS too small");
-		return -ENOMEM;
+	/* In case path level is set to 'none' or 'object' and we have only default oi */
+	path->level = LWM2M_PATH_LEVEL_OBJECT_INST;
+
+	ret = put_basename(out, path);
```

Comment:
we write base name here again?

Reply:
this state is ok

Reply:
Yeah, this is the only place where it needs to be written.